### PR TITLE
Upgradeable proxy example and refactor

### PR DESCRIPTION
**Pull request by wwhchung**
- adding prettier and solhint
- adding upgradeable proxy examples
- refactoring code generation and token/salt lookup into a library
- adding tests for registry account creation failure cases

### CODE REVIEW

File: src/lib/ERC6551AccountByteCode.sol [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            uint256,
+            address,
+            uint256
+        )
+    {
+        // codeAt start = creationCode.length-20
+        // codeAt end = creationCode.length+76
+        return
+            abi.decode(
+                codeAt(address(this), 151, 247),
```
[the remark this answers is not in the document]
Comment (wwhchung) [reply src/lib/ERC6551AccountByteCode.sol:88]: Isn't this more expensive though?  these functions are used commonly.  It's additional gas cost.

File: test/ERC6551Registry.t.sol [hunk 2]
```diff
@@ -6,6 +6,64 @@ import "forge-std/Test.sol";
...
+        MockERC6551Account accountInstance = MockERC6551Account(payable(deployedAccount));
+
+        (uint256 chainId_, address tokenAddress_, uint256 tokenId_) = accountInstance.token();
+        assertEq(chainId_, chainId);
+        assertEq(tokenAddress_, tokenAddress);
+        assertEq(tokenId_, tokenId);
+
+        assertEq(salt, accountInstance.salt());
+
+        console.log(accountInstance.codeLength());
```
[the remark this answers is not in the document]
Comment (wwhchung) [reply test/ERC6551Registry.t.sol:64]: that was a debug thing.  removing

File: test/mocks/MockERC6551Account.sol [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    function supportsInterface(bytes4 interfaceId) public view returns (bool) {
+        if (interfaceId == 0xffffffff) return false;
+        return _initialized;
+    }
+
+    function codeLength() external view returns (uint256) {
+        return ERC6551AccountByteCode.codeLength();
+    }
```
[the remark this answers is not in the document]
Comment (wwhchung) [reply test/mocks/MockERC6551Account.sol:45]: Ya, this was a debug thing, committed by accident, removing

File: .prettierrc.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
+  "printWidth": 100,
```
[the remark this answers is not in the document]
Comment (wwhchung) [reply .prettierrc.json:2]: removing

File: .prettierrc.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  "proseWrap": "preserve",
+  "overrides": [
+    {
+      "files": "*.sol",
+      "options": {
+        "printWidth": 100,
```
[the remark this answers is not in the document]
Comment (wwhchung) [reply .prettierrc.json:8]: removing

File: .solhint.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "solhint:recommended",
+  "rules": {
+    "compiler-version": ["error", "^0.8.0"],
+    "func-visibility": ["warn", { "ignoreConstructors": true }],
+    "avoid-low-level-calls": "off",
+    "no-inline-assembly": "off",
+    "avoid-tx-origin": "off"
```
[the remark this answers is not in the document]
Comment (wwhchung) [reply .solhint.json:8]: Removing

File: src/examples/upgradeable/ERC6551AccountUpgradeable.sol [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        bytes memory
+    ) public view returns (bytes4) {
+        (uint256 _chainId, address _contractAddress, uint256 _tokenId) = ERC6551AccountByteCode
+            .token();
+        require(
+            _chainId != block.chainid ||
+                msg.sender != _contractAddress ||
+                receivedTokenId != _tokenId,
+            "Cannot own yourself"
+        );
```
[the remark this answers is not in the document]
Comment (wwhchung) [reply src/examples/upgradeable/ERC6551AccountUpgradeable.sol:61]: Moved into _revertIfOwnershipCycle, but not sure how it blocks tokens for same collection.  All it's saying is that it requires at least one of chainId, token contract address or tokenId to NOT match.

File: src/examples/upgradeable/ERC6551AccountUpgradeable.sol [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        uint256 receivedTokenId
+    ) internal view {
+        address currentOwner = owner();
+        require(currentOwner != address(this), "Token in ownership chain");
+
+        uint32 currentOwnerSize;
+        /// @solidity memory-safe-assembly
+        assembly {
+            currentOwnerSize := extcodesize(currentOwner)
+        }
```
[the remark this answers is not in the document]
Comment (wwhchung) [reply src/examples/upgradeable/ERC6551AccountUpgradeable.sol:102]: Change being made

File: src/examples/upgradeable/ERC6551AccountUpgradeable.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (jaydenwindle) [reply src/examples/upgradeable/ERC6551AccountUpgradeable.sol]: ah my bad, so used to revert clauses

File: src/lib/ERC6551AccountByteCode.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaydenwindle) [reply src/lib/ERC6551AccountByteCode.sol:77]: that is true

File: .prettierrc.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
+{
+  "proseWrap": "preserve",
+  "printWidth": 100,
```
Comment (jaydenwindle) [inline .prettierrc.json:3]:
```suggestion
```

**Review by jaydenwindle — APPROVED** [review]

**Review by jaydenwindle — APPROVED** [review]

**Pull request merged**